### PR TITLE
[Op][NN1] Conv2D, MaxPool2D, AdaptiveAvgPool2D, ReLU, GeLU, SiLU, Softmax

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR introduces NN operators to the latest StructInfo codebase. It is part of our operator migration efforts as listed in https://github.com/mlc-ai/relax/issues/62.

Considering the PR length, to ease review, we split the migration of NN operators by multiple parts and this PR is the first part. It contains the following neural network operators.
* Convolution: Conv2D
* Pooling: MaxPool2D, AdaptiveAvgPool2D
* Linear unit: ReLU, GeLU, SiLU
* Softmax

This PR lifts some common code that is sharable by different kinds of operators to `op_common.h`.

### CODE REVIEW

File: src/relax/op/op_common.h [hunk 1]
```diff
@@ -33,8 +33,172 @@
...
+ * \param sinfo The input tensor struct info to be checked.
+ * \param expected_ndim The ndim that the input tensor struct info is expected to have
+ * \param op_name The name of the operator that invokes this function.
+ * \return The shape of the input tensor in ShapeExpr, or `NullOpt` if the shape is unknown.
+ */
+inline Optional<ShapeExpr> CheckNdimAndGetShape(const Call& call, const BlockBuilder& ctx,
+                                                const TensorStructInfo& sinfo, int expected_ndim,
+                                                const String& op_name) {
+  if (!sinfo->IsUnknownNdim() && sinfo->ndim != expected_ndim) {
+    ctx->ReportFatal(Diagnostic::Error(call->span)
```
Comment (tqchen) [inline src/relax/op/op_common.h]: in error message, avoid passing in call->span, instead directly pass in call. This would helps us to handle cases where call do not have a span

File: src/relax/op/op_common.h [hunk 2]
```diff
@@ -33,8 +33,172 @@
...
+      .add_argument("e", "Tensor", "The input tensor.")               \
+      .set_attr<FInferStructInfo>("FInferStructInfo", InferStructInfoUnary)
+
+inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx) {
+  return GetUnaryInputTensorStructInfo(call, ctx, /*op_name=*/"Unary");
+}
+
+/************ Utilities ************/
+
+using FComputeOutDtype = std::function<DataType(const Call&, const BlockBuilder&,
```
Comment (tqchen) [inline src/relax/op/op_common.h]: this seems to be local to InferShapeBinaryBroadcast. Also it is helpful to directly pass in FComputeOutDtype as a template parameter instead of std function. The lambda can be inlined by compiler.

Comment (tqchen) [reply src/relax/op/op_common.h]:
```c++
template<typename FType>
StructInfo InferStructInfoBroadcast(const Call&, BlockBuilder, FType fcompute_dtype) {
}
```

File: src/relax/op/nn/convolution.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  PrimExpr input_channel_kernel = _weight_shape->values[_kernel_layout.find('I')];
+  if (analyzer->CanProve(input_channel_data - input_channel_kernel != 0)) {
+    ctx->ReportFatal(Diagnostic::Error(call->span)
+                     << "The channel size of the data should equal to the input channel size of "
+                        "the weight. However, the data channel size is "
+                     << input_channel_data << " while the weight input channel size is "
+                     << input_channel_kernel);
+  }
+
+  std::unordered_map<char, PrimExpr> output_shape;
```
Comment (tqchen) [inline src/relax/op/nn/convolution.cc]: avoid build your own data layout handling logic. Instead reuse `tir/data_layout.h`


See examples here https://github.com/tlc-pack/relax/blob/relax/src/relay/op/nn/convolution.cc#L71

File: src/relax/op/nn/pooling.cc [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    return TensorStructInfo(data_sinfo->dtype, ndim);
+  }
+
+  const auto* attrs = call->attrs.as<MaxPool2DAttrs>();
+  ShapeExpr _data_shape = data_shape.value();
+  std::string _data_layout = attrs->layout.operator std::string();
+
+  std::unordered_map<char, PrimExpr> output_shape;
+  PrimExpr batch = _data_shape->values[_data_layout.find('N')];
+  PrimExpr channel = _data_shape->values[_data_layout.find('C')];
```
Comment (tqchen) [inline src/relax/op/nn/pooling.cc]: Use tir/data_layout.h mechanism

File: tests/python/relax/test_op_nn_convolution.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            ),
+            "float32",
+        ),
+    )
+
+
+def test_conv2d_infer_struct_info_more_input_dtype():
+    bb = relax.BlockBuilder()
+    x0 = relax.Var("x", R.Tensor((2, 3, 28, 28), "float16"))
+    w0 = relax.Var("w", R.Tensor((4, 3, 3, 3), "float16"))
```
Comment (tqchen) [inline tests/python/relax/test_op_nn_convolution.py:195]: Add a testcase of packed Layout. `NCHW16c`

File: tests/python/relax/test_op_nn_pooling.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+
+def _check_inference(bb: relax.BlockBuilder, call: relax.Call, expected_sinfo: relax.StructInfo):
+    ret = bb.normalize(call)
+    tvm.ir.assert_structural_equal(ret.struct_info, expected_sinfo)
+
+
+def test_max_pool2d_infer_struct_info():
+    bb = relax.BlockBuilder()
+    x0 = relax.Var("x", R.Tensor((2, 3, 32, 32), "float32"))
```
Comment (tqchen) [inline tests/python/relax/test_op_nn_pooling.py:39]: Add a case with symbolic input shape

File: tests/python/relax/test_op_nn_convolution.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    x1 = relax.Var("x", R.Tensor((2, 28, 28, 3), "float32"))
+    x2 = relax.Var("x", R.Tensor("float32", ndim=4))
+    x3 = relax.Var("x", R.Tensor("float32"))
+    x4 = relax.Var("x", R.Tensor())
+    x5 = relax.Var("x", R.Tensor((2, 4, 28, 28, 16), "float32"))
+    w0 = relax.Var("w", R.Tensor((4, 3, 3, 3), "float32"))
+    w1 = relax.Var("w", R.Tensor((3, 4, 3, 3), "float32"))
+    w2 = relax.Var("w", R.Tensor("float32", ndim=4))
+    w3 = relax.Var("w", R.Tensor("float32"))
+    w4 = relax.Var("w", R.Tensor((48, 4, 3, 3, 16), "float32"))
```
Comment (tqchen) [inline tests/python/relax/test_op_nn_convolution.py:49]: Add a case with symbolic input shape

**Comment by MasterJH5574** [comment]
@tqchen Added the support for layouts like “NCHW16c” and updated per your comments. `tir::Layout` is some utility that’s really helpful. Please kindly take another look :-)

**Review by tqchen — APPROVED** [review]
Thanks @MasterJH5574 please add a few cases with symbolic input shape for conv, pooling and I think we are good.

**Comment by tqchen** [comment]
Seems we already have then, then we are good
